core/src/main/java/org/mule/processor/AbstractRequestResponseMessageProcessor.java

```diff
@@ -0,0 +1,153 @@
+    {
+        final ReplyToHandler originalReplyToHandler = event.getReplyToHandler();
+        event = new DefaultMuleEvent(event, new CompletionHandlerReplyToHandlerAdaptor(new CompletionHandler<MuleEvent, MessagingException>()
+        {
+            @Override
+            public void onCompletion(MuleEvent event)
+            {
+                try
+                {
+                    MuleEvent response = processResponse(new DefaultMuleEvent(event, originalReplyToHandler));
+                    if (!NonBlockingVoidMuleEvent.getInstance().equals(response))
+                    {
+                        originalReplyToHandler.processReplyTo(response, null, null);
+                    }
+                    processFinallly(event, null);
+                }
+                catch (MuleException e)
+                {
+                    onFailure(new MessagingException(event, e));
+                }
+            }
+
+            @Override
+            public void onFailure(MessagingException exception)
+            {
+                originalReplyToHandler.processExceptionReplyTo(exception.getEvent(), exception, null);
+                processFinallly(exception.getEvent(), exception);
+            }
+        }));
+
+        try
+        {
+            MuleEvent result = processNext(processRequest(event));
+            if (!(result instanceof NonBlockingVoidMuleEvent))
+            {
+                return processResponse(event);
+            }
+            else
+            {
+                return result;
+            }
+        }
+        catch (MessagingException exception)
+        {
+            processFinallly(event, exception);
+            throw exception;
+        }
+    }
+
+    private boolean isNonBlocking(MuleEvent event)
+    {
+        return event.isAllowNonBlocking() && event.getReplyToHandler() != null && next != null;
+    }
+
+    /**
+     * Processes the request phase before the next message processor is invoked.
+     *
+     * @param event event to be processed.
+     * @return result of request processing.
+     * @throws MuleException exception thrown by implementations of this method whiile performing response processing
+     */
+    protected MuleEvent processRequest(MuleEvent event) throws MuleException
+    {
+        return event;
+    }
+
+    /**
+     * Processes the response phase after the next message processor and it's response phase have been invoked
+     *
+     * @param event event to be processed.
+     * @return result of response processing.
+     * @throws MuleException exception thrown by implementations of this method whiile performing response processing
+     */
+    protected MuleEvent processResponse(MuleEvent event) throws MuleException
+    {
+        return event;
+    }
+
+    /**
+     * Use to perform post processing after both request and response phased have been processed both in the case of a
+     * successful result and in the case of an exception being thrown.
```

Comment:
"both in the case of a successful result and in the case of an exception being thrown" seems redundant